multistream-2.0/spec.md

```diff
@@ -0,0 +1,403 @@
+   an issue as we can run as many sub-protocols over the same stream as we want.
+2. To reduce RTTs, all protocols are unidirectional.
+3. These protocols were *also* designed to eventually support packet protocols
+   (the other reason to be unidirectional and a strong motivator for the
+   `speculative-stream` and `serial-stream` protocols).
+
+### Multistream Advertise
+
+Unspeced (for now). Really, we just need to send a mapping of protocol
+names/codecs to contextual IDs (and may be some service discovery information).
+Basically, identify.
+
+### Multistream Use
+
+The `multistream/use` protocol is simply two varint multicodecs: the
+multistream-use multicodec followed by the multicodec for the protocol to be
+used. This protocol is *unidirectional*. If the stream is bidirectional, the
+receiver must acknowledge a successful protocol negotiation by responding with
+the same multistream-use protocol sequence.
+
+Every stream starts with multistream-use. Every other protocol defined here will
+be assigned a multicodec and selected with `multistream/use.`
+
+This protocol should *also* be trivial to optimize in hardware simply by prefix
+matching (i.e., matching on the first N (usually 16-32) bits of the
+stream/message).
+
+* [ ] Q: Technically, the first multicodec is redundant. However, it acts as a
+      magic byte that allows us to figure out what's going on. Should we keep
+      it? We could just start all streams with a single multicodec representing
+      the protocol
+* [ ] Q: Should we somehow distinguish between initiator and receiver? Should we
+      distinguish between bidirectional and unidirectional? We could even bit
+      pack these options into a single byte and use this instead of the leading
+      multicodec...
+
+### Multistream Dynamic
+
+The `multistream/dynamic` protocol is like the `multistream/use` protocol
+*except* that it uses a string to identify the protocol. To do so, the initiator
+simply sends a varint length followed by the name of the protocol.
+
+Including the `multistream/use` portion, the initiator would send:
+
+```
+<multistream/use><multistream/dynamic><length(varint)><name(string)>
+```
+
+Note: This used to use a fixed-width 16 bit number for a length. However, a
+varint *really* isn't going to cost us much, if anything, in terms of
+performance as most protocol names will be <= 128 bytes long. On the other hand,
+using different number formats everywhere *will* cost us in terms of complexity.
+
+### Multistream Contextual
+
+The `multistream/contextual` protocol is used to select a protocol using a
+*receiver specified*, session-ephemeral protocol ID. These IDs are analogues of
+ephemeral ports.
+
+In this protocol, the stream initiator sends a varint ID specified by the
+*receiver* to the receiver. This is a *unidirectional* protocol.
+
+Format:
+
+```
+<multistream/use><multistream/contextual><id(varint)>
+```
+
+The ID 0 is reserved for saying "same protocol" on a bidirectional stream. The
```

Comment:
We should probably just say "ID 0 means YES".

The issues are:

1. These protocols all have unidirectional variants.
2. Who initiated the stream isn't, unfortunately, always unambiguous.
3. These contextual IDs are *relative* and, even if you know my contextual ID for a protocol, I may not know yours.

Given these three issues:

1. I can't use *your* ID because I may not know it.
2. I can't use *mine* because your multistream muxer is expecting *your* IDs.

Therefore, I'm using 0.

However, there's probably a better way to say this.